// cdk-dia/dist/src/diagram/aws/stack-exports-container.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StackExportsContainer = void 0;
const uniqueid_1 = require("@aws-cdk/core/lib/private/uniqueid");
class StackExport {
    constructor(targetId, fullId) {
        this.targetId = targetId;
        this.fullId = fullId;
    }
}
class StackExportsContainer {
    constructor() {
        this.exports = [];
    }
    static fromComponent(exportsComponent) {
        const container = new StackExportsContainer();
        exportsComponent.children.forEach(exportNode => {
            const refMatches = exportNode.id.match(new RegExp('Output{"Ref":"(.*)"}'));
            if (refMatches) {
                const refTargetId = refMatches[1];
                container.exports.push(new StackExport(refTargetId, exportNode.id));
            }
            const getAttMatches = exportNode.id.match(new RegExp('Output{"Fn::GetAtt":\\["(.*)",".*"\\]}'));
            if (getAttMatches) {
                const refTargetId = getAttMatches[1];
                container.exports.push(new StackExport(refTargetId, exportNode.id));
            }
        });
        return container;
    }
    getExport(edgeTargetStackExport) {
        const stackExport = this.exports.find(stackExport => {
            const calculatedExportName = (0, uniqueid_1.makeUniqueId)(['Exports', stackExport.fullId]);
            return calculatedExportName == edgeTargetStackExport.exportValue;
        });
        return stackExport ? stackExport.targetId : undefined;
    }
}
exports.StackExportsContainer = StackExportsContainer;
