// cdk-dia/dist/src/cdk/cdk-tree.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Node = exports.ConstructInfoFqn = exports.ConstructInfoFqnType = exports.Tree = void 0;
class Tree {
    static fromObject(object) {
        const tree = new Tree();
        tree.version = object['version'];
        tree.tree = Node.fromObject(object['tree']);
        return tree;
    }
}
exports.Tree = Tree;
var ConstructInfoFqnType;
(function (ConstructInfoFqnType) {
    ConstructInfoFqnType[ConstructInfoFqnType["STACK"] = 0] = "STACK";
    ConstructInfoFqnType[ConstructInfoFqnType["STAGE"] = 1] = "STAGE";
    ConstructInfoFqnType[ConstructInfoFqnType["APP"] = 2] = "APP";
    ConstructInfoFqnType[ConstructInfoFqnType["CUSTOM_RESOURCE"] = 3] = "CUSTOM_RESOURCE";
})(ConstructInfoFqnType = exports.ConstructInfoFqnType || (exports.ConstructInfoFqnType = {}));
class ConstructInfoFqn {
    constructor(cdkType) {
        this.typ = cdkType;
    }
}
exports.ConstructInfoFqn = ConstructInfoFqn;
ConstructInfoFqn.of = (val) => {
    switch (val) {
        case "@aws-cdk/core.Stack":
        case "aws-cdk-lib.Stack":
            return new ConstructInfoFqn(ConstructInfoFqnType.STACK);
        case "@aws-cdk/core.Stage":
        case "aws-cdk-lib.Stage":
            return new ConstructInfoFqn(ConstructInfoFqnType.STAGE);
        case "@aws-cdk/core.App":
        case "aws-cdk-lib.App":
            return new ConstructInfoFqn(ConstructInfoFqnType.APP);
        case "@aws-cdk/core.CustomResource":
        case "aws-cdk-lib.CustomResource":
            return new ConstructInfoFqn(ConstructInfoFqnType.CUSTOM_RESOURCE);
        default:
            return undefined;
    }
};
class Node {
    constructor() {
        this.constructInfoFqn = undefined;
    }
    findInSubTree(predicate) {
        if (predicate(this))
            return this;
        const childrenArr = Array.from(this.children);
        for (const childrenKey in childrenArr) {
            const node = childrenArr[childrenKey][1].findInSubTree(predicate);
            if (node != null)
                return node;
        }
        return null;
    }
    static fromObject(object) {
        const node = new Node();
        node.id = object['id'];
        node.path = object['path'];
        node.children = new Map();
        if (object['children'] != undefined) {
            for (const childKey in object['children']) {
                node.children.set(childKey, Node.fromObject(object['children'][childKey]));
            }
        }
        node.attributes = new Map();
        if (object['attributes'] != undefined) {
            for (const attrKey in object['attributes']) {
                node.attributes.set(attrKey, object['attributes'][attrKey]);
            }
        }
        if (object['constructInfo'] != undefined && object['constructInfo']['fqn'] != undefined) {
            node.constructInfoFqn = ConstructInfoFqn.of(object['constructInfo']['fqn']);
        }
        return node;
    }
}
exports.Node = Node;
